### PR TITLE
Login page and auth context

### PR DESCRIPTION
**Pull request by jkerr0**
Added login page and auth context

### CODE REVIEW

File: frontend/src/app/login/page.tsx [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+
+    if (!validateRequired(login)) {
+      valid = false;
+      setLoginError('Field is required');
+    }
+    if (!validateRequired(password)) {
+      valid = false;
+      setPasswordError('Field is required');
+    }
```
Comment (Alschn) [inline frontend/src/app/login/page.tsx]: Use zod schema as we already have it installed

File: frontend/src/app/login/page.tsx [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+    if (valid) {
+      doLogin(login as string)
+      router.push('/');
+    }
+  }
+
+  return <>
+    <main className={cn('flex', 'items-center', 'justify-center', 'flex-col', 'min-h-screen')}>
```
Comment (Alschn) [inline frontend/src/app/login/page.tsx]: `cn` is only needed if there are some dynamic tailwind classes to be merged with existing styles. Simply pass className as one string instead.

File: frontend/src/components/UserLoginIndicator.tsx [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+import {Button} from "~/components/ui/button";
+import Link from "next/link";
+
+export default function UserLoginIndicator() {
+  const {login, logout} = useAuthContext()
+  return (
+    login
+      ? <span><span className={'px-2'}>{`Hello, ${login}`}</span><Button onClick={logout}>Logout</Button></span>
+      : <Link href={'/login'}><Button>Login</Button></Link>
+  )
```
Comment (Alschn) [inline frontend/src/components/UserLoginIndicator.tsx]: Imo *if statement* would be a little bit more readable than ternary in this case.

```react
if (!login) return (
  ...
)

return ...
```

File: frontend/src/components/ui/field.tsx [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+  error?: string;
+}
+
+export const Field = forwardRef<HTMLInputElement, InputProps>( function Field({error, label, ...props}: InputProps, ref) {
+  return (<div className={cn('flex', 'flex-col', 'gap-1')}>
+    <label htmlFor={props.id} className={cn('p-1')}>{`${label}:`}</label>
+    <input {...props} ref={ref} className={cn('rounded-md', 'p-2', 'border')}/>
+    {error && <span className={cn('dark:text-red-400','text-red-500', 'text-xs', 'py-2', 'px-1')}>{error}</span>}
+  </div>);
+})
```
Comment (Alschn) [inline frontend/src/components/ui/field.tsx]: Same issue with `cn` as previously. 

In this case, if Field component took `className` as prop, it would be justified to use `cn` there.
`<div className={cn("flex flex-col gap-1", className}>...</div>`

File: frontend/src/components/ui/form.tsx [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+export default function Form({onSubmit, fields, submitButtonLabel, ...props}: FormProps) {
+  const handleSubmit: FormEventHandler<HTMLFormElement> = e => {
+    e.preventDefault();
+    onSubmit();
+  }
+  return (
+    <div className={cn('border', 'rounded-md', 'w-1/3', 'p-3')}>
+      <form onSubmit={handleSubmit} {...props}>
+        {fields}
```
Comment (Alschn) [inline frontend/src/components/ui/form.tsx]: I guess it is debatable but imo `children` prop would fit there more than a "fields slot". If you happened to build components like that, then leave it as it is.

File: frontend/src/store/auth-context-provider.tsx [hunk 6]
```diff
@@ -0,0 +1,35 @@
+import {ReactNode, useReducer} from "react";
+import {AuthContext, AuthContextProps} from "./auth-context";
+
+enum ActionType {
+  LOGIN = 'login',
+  LOGOUT = 'logout'
+}
```
Comment (Alschn) [inline frontend/src/store/auth-context-provider.tsx]: You should (almost) always prefer objects marked with `as const` to Typescript's enum. In this case it does not really matter because action type is not used outside of this file and everything works as expected. However, if we were to use an enum in multiple cases in the project, we could stumble upon weird behaviours and typing issues.

Have a read, if you have not known about it before:

https://www.crocoder.dev/blog/typescript-enums-good-bad-and-ugly/

https://graphqleditor.com/blog/enums-are-still-bad/

tldr: typescript enum bad, use as const

```typescript
const ActionTypes = {
  LOGIN: "login",
  LOGOUT: "logout"
} as const;

type ActionType = typeof ActionTypes[keyof typeof ActionTypes];
```

File: frontend/src/store/auth-context-provider.tsx [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+  const reducer: AuthReducer = (state, action): AuthContextProps => {
+    switch (action.type) {
+      case ActionType.LOGIN:
+        return {...state, userLogin: action.payload}
+      case ActionType.LOGOUT:
+        return {...state, userLogin: undefined}
+      default:
+        return state
+    }
+  }
```
Comment (Alschn) [inline frontend/src/store/auth-context-provider.tsx]: Move reducer function definition outside of the component. This way it will be possible to test that function and make it "framework agnostic".

File: frontend/src/store/auth-context.tsx [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+
+export interface AuthContextProps {
+  userLogin: string | undefined;
+  setUserLogin: (login: string) => void;
+  logout: () => void;
+}
+
+const authContextDefaultValue: AuthContextProps = {
+  userLogin: undefined,
+  setUserLogin: () => {},
```
Comment (Alschn) [inline frontend/src/store/auth-context.tsx]: Imo auth context should store `user` object, `login` and `logout` functions. Authentication part is subject to change, so we will see how things develop.

File: frontend/src/store/use-auth-context.tsx [hunk 9]
```diff
@@ -0,0 +1,10 @@
+"use client"
+
+import { useContext } from "react";
+import {AuthContext} from "~/store/auth-context";
+
+export default function useAuthContext() {
+  const {userLogin, setUserLogin, logout} = useContext(AuthContext)
+
+  return {login: userLogin, doLogin: setUserLogin, logout}
+}
```
Comment (Alschn) [inline frontend/src/store/use-auth-context.tsx]: I would rethink what is to be stored in auth store and let `useAuthContext` simply work as an shortcut to `useContext(AuthContext)`. Extra remapping or adding logic that does not leave in auth provider seems a bit unnecessary.

Moreover usage of `login` and `logout` in different contexts (one is a string, another one is a function) is a bit misleading. Imo `username` fits much better to describe the field that user logged in with (even if its value was an email or a phone number).

File: frontend/src/components/ui/field.tsx [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+import {useTheme} from "next-themes";
+
+type HtmlInputProps = ComponentPropsWithoutRef<'input'>
+
+interface InputProps extends HtmlInputProps {
+  label: string;
+  error?: string;
+}
+
+export const Field = forwardRef<HTMLInputElement, InputProps>( function Field({error, label, ...props}: InputProps, ref) {
```
Comment (Alschn) [inline frontend/src/components/ui/field.tsx]: Btw Shadcn has some 🔥 components like Input and Form that take care of that boilerplate for you

**Review by Alschn — CHANGES_REQUESTED** [review]
Please fix some of the highlighted issues and run a formatter on files to keep consistency with existing code base (later we can add a precommit hook or something fancy like that to, to make this project even more spicier xD). If you do not agree with some of opinions of mine, lmk.

File: frontend/src/app/login/page.tsx [hunk 11]
```diff
@@ -4,9 +4,14 @@ import React, {createRef, useState} from "react";
 import {Field} from "~/components/ui/field";
 import {cn} from "~/lib/utils";
 import Form from "~/components/ui/form";
-import {validateRequired} from "~/lib/validations";
 import {useRouter} from "next/navigation";
 import useAuthContext from "~/store/use-auth-context";
+import {object, string, ZodError} from "zod";
```
Comment (Alschn) [inline frontend/src/app/login/page.tsx]: `import * as z from "zod"` and use `z.object` etc. because otherwise you are shadowing typescript's `object` and `string`.

File: frontend/src/app/login/page.tsx [hunk 12]
```diff
@@ -18,25 +23,32 @@ export default function LoginPage() {
...
-    let valid = true;
+  const handleValidationError = (error: ZodError) => {
+    error?.errors.forEach(({path, message}) => {
+      switch (path.toString()) {
+        case 'login':
+          setLoginError(message);
+          break;
+        case 'password':
+          setPasswordError(message);
+      }
```
Comment (Alschn) [inline frontend/src/app/login/page.tsx]: Missing `break` in password case

**Review by Alschn — CHANGES_REQUESTED** [review]

File: frontend/src/app/login/page.tsx (hunk 1 above)
Comment (jkerr0) [reply frontend/src/app/login/page.tsx]: Done

File: frontend/src/app/login/page.tsx (hunk 12 above)
Comment (jkerr0) [reply frontend/src/app/login/page.tsx]: Done

File: frontend/src/app/login/page.tsx (hunk 11 above)
Comment (jkerr0) [reply frontend/src/app/login/page.tsx]: Done

File: frontend/src/store/auth-context-provider.tsx (hunk 6 above)
Comment (jkerr0) [reply frontend/src/store/auth-context-provider.tsx]: I've heard that TS enums are bad, but never encountered any problems.

keyof typeof looks kinda funny, but i am open to learn some of TS magic 👍

File: frontend/src/components/ui/field.tsx (hunk 4 above)
Comment (jkerr0) [reply frontend/src/components/ui/field.tsx]: Ok, i've never used `cn` helper before.
That's my second time with tailwind in next. Before that i've only used it once to sprinkle my Go+HTMX tic-tac-toe mock project.

File: frontend/src/components/UserLoginIndicator.tsx (hunk 3 above)
Comment (jkerr0) [reply frontend/src/components/UserLoginIndicator.tsx]: Sure

File: frontend/src/app/login/page.tsx (hunk 2 above)
Comment (jkerr0) [reply frontend/src/app/login/page.tsx]: Done

File: frontend/src/store/use-auth-context.tsx (hunk 9 above)
Comment (jkerr0) [reply frontend/src/store/use-auth-context.tsx]: ok this is unnecessary

File: frontend/src/components/ui/field.tsx (hunk 10 above)
Comment (jkerr0) [reply frontend/src/components/ui/field.tsx]: I'll look more into shadcn

File: frontend/src/components/ui/form.tsx (hunk 5 above)
Comment (jkerr0) [reply frontend/src/components/ui/form.tsx]: "children" looks much nicer, yeah. i thought i'd make separate props for fields and buttons, but i've stuck somewhere in between and that was the result

**Review by Alschn — APPROVED** [review]
lgtm
